---
title: "Schraivogel (2020) Data Documentation"
author: "Eugene Katsevich"
date: "March 16, 2022"
output: github_document 
---

```{r, echo = FALSE, message = FALSE, warning = FALSE}
library(tidyverse)
```

# Overview 

The portion of the Schraivogel data that is currently imported is the one described in the section "TAP-seq sensitively detects gene expression changes" of their paper. It contains two separate experiments: one TAP-seq and one perturb-seq. These experiments are meant as a proof-of-concept for TAP-seq, so they contain only positive and negative control perturbations (perturbations for which the ground truth is known). Both experiments have the same experimental design; they differ only in that TAP-seq targets a small number of genes while perturb-seq targets the whole transcriptome.

The `processed` directory contains subdirectories corresponding to the two experiments:
```{r}
# top-level data directory
schraivogel_dir <-.get_config_path("LOCAL_SCHRAIVOGEL_2020_DATA_DIR")

# processed data directory
processed_dir <- sprintf("%sprocessed", schraivogel_dir)

# print subdirectories of `processed_dir`
list.dirs(processed_dir, full.names = FALSE, recursive = FALSE)
```
Each of these has subdirectories for the processed gRNA data, the processed gene expression data, and auxiliary data:
```{r}
# print subdirectories for one of the processed experiments
list.dirs(sprintf("%s/ground_truth_tapseq", processed_dir), 
          full.names = FALSE, recursive = FALSE)
```

# Experimental design

As mentioned above, the experimental design for these two ground truth experiments is the same. The file containing the experimental design is in the `aux` directory; let's take a look at its first few rows:

```{r}
aux_dir <- sprintf("%s/ground_truth_tapseq/aux", processed_dir)
exper_design = readRDS(sprintf("%s/experimental_design.rds", aux_dir))
exper_design %>% 
  head(5) %>% 
  kableExtra::kable(format = "html", 
                    booktabs = TRUE, 
                    col.names = c("gRNA", "Target", "Target Type", "Known Effect"))
```

The full table contains a total of `r nrow(exper_design)` rows, so there are `r nrow(exper_design)` gRNAs in this experiment. Breaking these down by their target,
```{r}
exper_design %>% pull(target) %>% table()
```
we see that we have `r exper_design %>% summarise(sum(target == "non-targeting")) %>% pull()` non-targeting gRNAs as well as four gRNAs each for 14 targets ($30 + 4\times14 = 86$). Of these targets, 10 are gene TSSs and 4 are well-characterized enhancers of four genes. 

# TAP-seq data

Let's now take a look at the TAP-seq data:
```{r}
# load the gRNA expression data
processed_gRNA_dir <- sprintf("%s/ground_truth_tapseq/gRNA", processed_dir)
gRNA_odm_fp <- sprintf("%s/raw_ungrouped.odm", processed_gRNA_dir)
gRNA_metadata_fp <- sprintf("%s/raw_ungrouped_metadata.rds", processed_gRNA_dir)
gRNA_expr_odm <- ondisc::read_odm(gRNA_odm_fp, gRNA_metadata_fp)
gRNA_expr_odm

# load the gene expression data
processed_gene_dir <- sprintf("%s/ground_truth_tapseq/gene", processed_dir)
gene_odm_fp <- sprintf("%s/expression_matrix.odm", processed_gene_dir)
gene_metadata_fp <- sprintf("%s/metadata.rds", processed_gene_dir)
gene_expr_odm <- ondisc::read_odm(gene_odm_fp, gene_metadata_fp)
gene_expr_odm
```
This experiment has `r ondisc::ncol(gene_expr_odm)` cells across `r gene_expr_odm %>% ondisc::get_cell_covariates() %>% summarise(length(unique(batch))) %>% pull()` batches. The gRNA data come in the form of expressions and are not thresholded. There are a total of `r ondisc::nrow(gRNA_expr_odm)` gRNAs, as discussed above. A total of `r ondisc::nrow(gene_expr_odm)` genes are measured. Based on the paper, there are supposed to be 74 genes measured: 14 that were targeted and 60 presumably unrelated genes. Of the two missing genes, one is HS2 (whose enhancer was targeted) and one is a presumably unrelated gene. We can look further into why this is the case, but perhaps it's not urgent.

# Perturb-seq data

Finally, we turn to the perturb-seq data:
```{r}
# load the gRNA expression data
processed_gRNA_dir <- sprintf("%s/ground_truth_perturbseq/gRNA", processed_dir)
gRNA_odm_fp <- sprintf("%s/raw_ungrouped.odm", processed_gRNA_dir)
gRNA_metadata_fp <- sprintf("%s/raw_ungrouped_metadata.rds", processed_gRNA_dir)
gRNA_expr_odm <- ondisc::read_odm(gRNA_odm_fp, gRNA_metadata_fp)
gRNA_expr_odm

# load the gene expression data
processed_gene_dir <- sprintf("%s/ground_truth_perturbseq/gene", processed_dir)
gene_odm_fp <- sprintf("%s/expression_matrix.odm", processed_gene_dir)
gene_metadata_fp <- sprintf("%s/metadata.rds", processed_gene_dir)
gene_expr_odm <- ondisc::read_odm(gene_odm_fp, gene_metadata_fp)
gene_expr_odm
```

This experiment has `r ondisc::ncol(gene_expr_odm)` cells across `r gene_expr_odm %>% ondisc::get_cell_covariates() %>% summarise(length(unique(batch))) %>% pull()` batches. The gRNA data come in the form of expressions and are not thresholded. There are a total of `r ondisc::nrow(gRNA_expr_odm)` gRNAs, which is one fewer than the 86 in the experimental design. Perhaps the missing one (`r setdiff(exper_design$gRNA_id, gRNA_expr_odm %>% ondisc::get_feature_ids())`) got removed during QC by Schraivogel et al? I am not sure. Unlike the TAP-seq experiment, we have measured the whole transcriptome (a total of `r ondisc::nrow(gene_expr_odm)` genes).
